src/qiskit_qasm3_import/converter.py

```diff
@@ -88,15 +93,45 @@
 }
 
 
+class AddressingMode:
+    """Addressing mode for qubits in OpenQASM 3 programs.
+
+    This class is useful as long as we allow only physical or virtual addressing modes, but
+    not mixed. If the latter is supported in the future, this class will be modified or removed.
+    """
+
+    def __init__(self):
+        self._state = 0  # 0 == UNKNOWN, 1 == PHYSICAL, 2 == VIRTUAL
+
+    def set_physical_mode(self, node):
+        """Set the addressing mode to physical. On success return `True`, otherwise `False`."""
+        if self._state == 2:
+            raise_from_node(
+                node,
+                "Physical qubit referenced in virtual addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = 1
+
+    def set_virtual_mode(self, node):
+        """Set the addressing mode to virtual. On success return `True`, otherwise `False`."""
+        if self._state == 1:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = 2
+
+
 class State:
-    __slots__ = ("scope", "source", "circuit", "symbol_table", "_unique")
+    __slots__ = ("scope", "source", "circuit", "symbol_table", "_unique", "addressing_mode")
```

Comment:
The addressing mode is a property of the _overall_ state of the parser, not just the state within a frame on the evaluation stack.  The addressing mode is purely a global-state thing, so the current mode must be accessible to `ConversionVisitor` at the point it's popping things off the stack.  I don't think this current form with local encapsulation will work with hardware qubits first encountered within a local scope.

I still don't like the class-based encapsulation, but I will go along with some form of it around the mode if you really prefer it.  However, I will insist that we use an `enum` for the mode - a comment explaining the values is not a substitute for correct typing.

Reply:
> The addressing mode is purely a global-state thing

Yep this is clear.

> I don't think this current form with local encapsulation will work with hardware qubits

I don't think so either. I didn't worry about whether to copy or reference the addressing mode in the construction of a the `State` for local scope. Because either way its broken. 

> That's not a clumsy manipulation, it's just following the rules of the language.

Yes, I realize that local scope and global scope are not the same thing. I wasn't clear about what I think is clumsy. The example I had in mind is that we search the entire symbol table at the end to see if there is a hardware qubit in it. I think the design needs to change a bit. Two symbol tables is just one idea. I did check how many places would have to search both of them, and it looked ok. But even storing the hardware qubits in a separate structure in `State` is not enough.

I considered an enum for the number labeling the mode as well, probably even prefer it. I omitted it because with a class wrapping the enum the code might start to look wordy. Plus the magic numbers are restricted to this one place in the code, which makes them less of an issue.